lib/evil_query.rb

```diff
@@ -63,7 +65,10 @@ def make_relation!
       logger.debug("[EVIL-QUERY] make_relation!")
       post_ids = aspects_post_ids! + ids!(followed_tags_posts!) + ids!(mentioned_posts)
       post_ids += ids!(community_spotlight_posts!) if @include_spotlight
-      Post.where(:id => post_ids)
+      logger.debug("Make blocked query")
+      Post.left_outer_joins(author: [:pod])
+          .where(id: post_ids)
+          .where("(pods.blocked = false or pods.blocked is null)")
```

Comment:
So here's the hard part about this :/ Our main stream query already is notoriously complex and thus slow. A lot of optimization went into it already in order to make it not too terribly slow. Extending it needs to be done very considerate and carefully, ideally doing some performance benchmarking on it, but at least comparing query plans before and after.

Any of that happened here? :)

Reply:
Actually, why do we register posts from blocked pod at all? We could just drop them. That would mean no way back of course but it would be much simpler.

Reply:
@jhass I checked the queries: 
Normal public posts query looks like this: 
```
SELECT  "posts".* FROM "posts" 
    WHERE "posts"."public" = true 
        AND (posts.created_at < '2021-04-11 17:49:52') 
        AND "posts"."type" IN ('StatusMessage', 'Reshare') AND (posts.author_id NOT IN (6,2)) 
    ORDER BY posts.created_at DESC, posts.id DESC LIMIT 15  
```

This results in this explanation (unsuprisingly): 
<img width="441" alt="Bildschirmfoto 2021-04-11 um 20 13 45" src="https://user-images.githubusercontent.com/501326/114316111-6d3ebc80-9b02-11eb-8ae7-bf8d1d66042d.png">

Query for Blocked posts: 
```
SELECT  "posts".* FROM "posts" 
    LEFT OUTER JOIN "people" ON "people"."id" = "posts"."author_id" 
    LEFT OUTER JOIN "pods" ON "pods"."id" = "people"."pod_id" 
    WHERE (pods.blocked is null or pods.blocked = false) AND "posts"."public" = true 
        AND (posts.created_at < '2021-04-11 17:51:47') 
        AND "posts"."type" IN ('StatusMessage', 'Reshare') AND (posts.author_id NOT IN (6,2)) 
    ORDER BY posts.created_at DESC, posts.id DESC LIMIT 15 
```
Explain: 
<img width="803" alt="Bildschirmfoto 2021-04-11 um 20 20 22" src="https://user-images.githubusercontent.com/501326/114316372-8431de80-9b03-11eb-8db7-8357e5de2c69.png">


Its, well-  more, but indexed. Is this a _real_ problem? 
The real bottleneck and time-consuming is IMHO the tons of extra queries done after this, beginning with 'like_posts_for_stream', Photos, Mentions, polls and so on. (Likes could also be joined to the posts query)

Reply:
@Flaburgan 
> Actually, why do we register posts from blocked pod at all? We could just drop them. That would mean no way back of course but it would be much simpler.

That was also my first Idea, but consider: 
A Podmin might be retain to require the data for legal reasons. (might be an extra function)
A Podmin might only temporary block - and unblock a pod after more investigation.
A Podmins can track what is currently blocked. 

My idea was not to delete any data.

Reply:
If a pod gets blocked the messages should not only be dropped on federation level, it should also be rejected/retracted for relayables like when the user is blocked. Otherwise users would federate content on their posts which they might want to delete.

Also as @jhass said, the stream query is already complex enough and was a reason for slow streams multiple times, so lets not make that more complex than it needs to be. And filtering/rejecting at federation level would remove the need of adding more stuff to the stream query.

Reply:
I got stuck stopping relay messages complete, here my knowledge about diaspora internal comes to an end. 

@SuperTux88 : Do you mean, to delete all posts( and user and comments)  from foreign Pods if this gets blocked? Thats quite a hard step. If I get you right, should we consider to 'archive' these deleted posts somewhere? No to recover this, but to have a copy for later. - maybe for legal reasons? 

What do you think about the Query explanation?

Reply:
Comments/likes/poll are relayed through the authors pod, if you hide them from the author, the blocked pod could still spam on their posts, but the author can do nothing against it which would be really bad. So if a pod is blocked, these messages need to be properly rejected (and not just dropped) and especially not just silently being forwarded to other pods and not telling the post author about it.

And you don't need to delete existing entities when blocking a pod (can be an optional step, if a podmin only wants to block future messages or also delete past stuff), but if it's in the database, it should also be shown to the users to give them the options to moderate their posts. So only filtering on incoming message, **no filtering at displaying** (and therefore also no need to change that query).

If you want to still store stuff for legal reasons take a backup or a screenshot or whatever you need.

Reply:
@SuperTux88 
In the meantime I have blown up my developer Database to 4Million posts and I see the problem.  I agree to reject any (future) requests to this pods and not changing the posts query. (So old entries are still visible).
Has this to be done in the federation-code or still on the diaspora site? Starting points for me? 

Another function (or job) may be manual triggered or automatic scheduled to clean up posts, comments etc from such blocked pods.